### PR TITLE
Implement unconditional CC encryption 

### PR DESCRIPTION
**Pull request by abstratt**
Replaces env vars with keystore as source of encryption keys.
Unconditionally encrypt the CC state files

Fixes: #23866

### CODE REVIEW

**Comment by abstratt** [comment]
@bot-gradle test APT

**Comment by bamboo** [comment]
🤔 It seems the encryption key hash should be part of the configuration cache key for the case someone deletes `$GRADLE_USER_HOME/gradle.keystore`.

**Comment by abstratt** [comment]
> 🤔 It seems the encryption key hash should be part of the configuration cache key for the case someone deletes `$GRADLE_USER_HOME/gradle.keystore`.

It is, right? Also, does this test cover that case?

https://github.com/gradle/gradle/pull/24243/files#diff-406b5abd6fdd148ae586d78a9f2d60819f4e7dce3481b9cc3e605538c65603d7R161

**Comment by abstratt** [comment]
I expect performance test failures here. I will do the rebaseline with the latest commit as soon as that happens.

**Comment by abstratt** [comment]
@bot-gradle test and merge

**Comment by bamboo** [comment]
> You don't see it in this changeset because it is already like that in release

Thanks, I missed that.

**Review by bamboo — APPROVED** [review]

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheEncryptionIntegrationTest.groovy [hunk 1]
```diff
@@ -160,109 +143,96 @@ class ConfigurationCacheEncryptionIntegrationTest extends AbstractConfigurationC
...
-        runWithEncryption(false)
+        keyStorePassword = "foobar"
+        fails(*(["help", "--configuration-cache"] + encryptionOptions))
 
         then:
-        configurationCache.assertStateStored()
-        outputContains("Calculating task graph as no configuration cache is available for tasks: help")
+        result.assertHasErrorOutput """
+Error loading encryption key from Java keystore at ${keyStorePath}
+> Integrity check failed: java.security.UnrecoverableKeyException: Failed PKCS12 integrity checking
```
Comment (bamboo) [inline subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheEncryptionIntegrationTest.groovy:157]: ❓ What should users do in this case?

**Comment by mlopatkin** [comment]
Doesn't relying on a keystore only complicate the ephemeral CI case a bit? Suppose I have a build with Github Actions using the Gradle's action. The action backs up the configuration cache directory and some stuff in `GRADLE_USER_HOME`.

How do you envision keystore management in this case? We cannot cache the keystore with the default password, as it defeats the purpose. I see two options, neither is appealing to me:
1. Add a step to generate the keystore manually from the environment variable
2. Add a step to create `gradle.properties` with the keystore password

AFAIK, using a command-line argument to pass passwords (`-Dorg.gradle.configuration-cache.internal.key-store-password=$MY_PASSWORD`) is frowned upon, as command-lines are widely readable.

To clarify, my threat model includes attacker being able to get read access to caches, but not action secrets - I hope Github stores them separately.

**Comment by mlopatkin** [comment]
> Doesn't relying on a keystore only complicate the ephemeral CI case a bit?

Ok, now I get to read the whole discussion.

**Comment by abstratt** [comment]
@mlopatkin The use of internal system properties at the moment are just to make things more easily testable. Same for enabling/disabling encryption, that will not be documented, but helps confirm things leak as plain contents without encryption, and that if some scenario is failing with encryption, but not without encryption, encryption is to blame.

Re: how to source a password to encrypt the keystore itself, you are right that still needs to be handled, and I wanted to address it separately, as a follow-up PR.

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheEncryptionIntegrationTest.groovy (hunk 1 above)
Comment (abstratt) [reply subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheEncryptionIntegrationTest.groovy:157]: Deleting the keystore would make it go away

**Comment by abstratt** [comment]
@bot-gradle test and merge

**Comment by abstratt** [comment]
@bot-gradle test and merge

**Review by hythloda — APPROVED** [review]

**Comment by bamboo** [comment]
@bot-gradle test ACC

**Comment by bamboo** [comment]
@bot-gradle test and merge

**Comment by abstratt** [comment]
@bot-gradle test and merge

**Comment by abstratt** [comment]
@bot-gradle test and merge

**Comment by ljacomet** [comment]
I have changed the target branch and milestone to reflect we are pushing this to Gradle 8.2

**Comment by bamboo** [comment]
Superseded by #24293

**Pull request closed without merging**